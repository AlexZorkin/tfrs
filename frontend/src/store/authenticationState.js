import { put, takeLatest, all } from 'redux-saga/effects'
import { getLoggedInUser } from '../actions/userActions'
import { initKeycloak, loginKeycloakUserSuccess } from '../actions/keycloakActions'
import Keycloak from 'keycloak-js'
import ActionTypes from '../constants/actionTypes/Keycloak'
import configureAxios from './authorizationInterceptor'
import CONFIG from '../config'

export default function * authenticationStateSaga (store) {
  const { token, expiry } = store.getState().userAuth
  const now = Math.round(Date.now() / 1000)
  // Token already exists so setup things here
  if (token && now < expiry) {
    configureAxios()
    yield put(getLoggedInUser())
  }

  const keycloak = new Keycloak({
    url: CONFIG.KEYCLOAK.AUTH_URL,
    realm: CONFIG.KEYCLOAK.REALM,
    clientId: CONFIG.KEYCLOAK.CLIENT_ID
  })
  const authenticated = yield keycloak.init({
    pkceMethod: 'S256',
    redirectUri: CONFIG.KEYCLOAK.CALLBACK_URL,
    idpHint: 'idir'
  })

  yield put(initKeycloak(keycloak, authenticated))

  yield all([
    takeLatest(ActionTypes.LOGIN_KEYCLOAK_USER_SUCCESS, getBackendUser)
  ])

  if (authenticated) {
    yield put(loginKeycloakUserSuccess(keycloak.idToken, keycloak.idTokenParsed.exp))
  }
}

function * getBackendUser (action) {
  configureAxios(action.payload)
  yield put(getLoggedInUser())
}
