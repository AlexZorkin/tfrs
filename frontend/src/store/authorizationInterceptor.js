import axios from 'axios'
// import userManager from './oidc-usermanager';
import CONFIG from '../config'
import { keycloak } from '../actions/keycloakActions'

function configureAxios () {
  if (CONFIG.KEYCLOAK) {
    axios.interceptors.request.use((config) => {
      const loadconfig = async () => {
        // const user = await userManager.getUser();
        const kc = keycloak() // TODO see if this needs to be an async get
        // console.log("configureAxios user", kc)
        return {
          ...config,
          headers: {
            Authorization: `Bearer ${kc.idToken}`
          }
        }
      }

      return loadconfig()
    })
  }
}

export default configureAxios
