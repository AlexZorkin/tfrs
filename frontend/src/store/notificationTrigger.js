import { put, takeLatest, delay } from 'redux-saga/effects'
import { getNotifications, getNotificationsCount } from '../actions/notificationActions'
import UserActionTypes from '../constants/actionTypes/Users'
import NotificationActionTypes from '../constants/actionTypes/Notifications'

const TRIGGERING_ACTIONS = [
  'SERVER_INITIATED_NOTIFICATION_RELOAD',
  UserActionTypes.RECEIVE_LOGGED_IN_USER,
  NotificationActionTypes.SUCCESS_NOTIFICATIONS
]

function * fetchNotifications (store) {
  yield delay(1000) // debounce
  if (store.getState().rootReducer.userRequest.isAuthenticated &&
        store.getState().rootReducer.userRequest.loggedInUser?.id) {
    yield put(getNotificationsCount())
  }

  if (store.getState().rootReducer.notifications.onNotificationsPage) {
    yield put(getNotifications())
  }
}

export default function * notificationsSaga (store) {
  yield takeLatest(
    action => (TRIGGERING_ACTIONS.includes(action.type)),
    fetchNotifications,
    store
  )
}
