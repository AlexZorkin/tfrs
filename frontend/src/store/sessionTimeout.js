import { put, takeLatest, delay } from 'redux-saga/effects'
import { signUserOut } from '../actions/userActions'
import ActionTypes from '../constants/actionTypes/Keycloak'

const NON_RESETTING_ACTIONS = [
  'RECEIVE_NOTIFICATIONS',
  'RECEIVE_NOTIFICATIONS_COUNT',
  'GET_NOTIFICATIONS_COUNT',
  'SERVER_INITIATED_CONNECTION_RELOAD',
  ActionTypes.LOGIN_KEYCLOAK_USER_SUCCESS,
  'redux-oidc/USER_EXPIRING', // TODO Handle user expired via keycloak
  'SESSION_TIMEOUT_RESET',
  'SESSION_TIMEOUT_WARNING',
  'SESSION_TIMEOUT_EXPIRED'
]

function * resetTimer () {
  console.log('resetting timer')
  yield put({ type: 'SESSION_TIMEOUT_RESET' })
  yield delay(10 * 60 * 1000)
  yield put({ type: 'SESSION_TIMEOUT_WARNING' })
  yield delay(3 * 60 * 1000)
  yield put({ type: 'SESSION_TIMEOUT_EXPIRED' })
  yield put(signUserOut())
}

export default function * sessionTimeoutSaga () {
  yield takeLatest(
    action => (!NON_RESETTING_ACTIONS.includes(action.type)),
    resetTimer
  )
}
