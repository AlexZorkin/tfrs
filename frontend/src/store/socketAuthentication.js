import { put, takeLatest, all } from 'redux-saga/effects'
import { getKeycloakUser } from '../actions/keycloakActions'
import ActionTypes from '../constants/actionTypes/Keycloak'

function socketAuthenticate (dispatch) {
  console.log('socketAuthenticate')

  const user = getKeycloakUser() // TODO see if this needs to be an asycn get

  if (!user || !user.id_token) { return }

  dispatch({
    type: 'socketio/AUTHENTICATE',
    token: user.id_token
  })
}

function * socketDeauthenticate () {
  yield put({
    type: 'socketio/DEAUTHENTICATE'
  })
}

export default function * socketAuthenticationSaga (store) {
  yield all([
    takeLatest(ActionTypes.LOGIN_KEYCLOAK_USER_SUCCESS, socketAuthenticate, store.dispatch),
    takeLatest('redux-oidc/USER_EXPIRED', socketDeauthenticate) // TODO Handle user expired via keycloak
  ])
}
