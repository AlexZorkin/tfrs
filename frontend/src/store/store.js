import createSagaMiddleware from 'redux-saga'
import { compose } from 'redux'
import thunk from 'redux-thunk'
import { configureStore } from '@reduxjs/toolkit'
import persistState from 'redux-localstorage'
import { createLogger } from 'redux-logger'
import createSocketIoMiddleware from 'redux-socket.io'
import { reducer as toastrReducer } from 'react-redux-toastr'
import io from 'socket.io-client'

import rootReducer from '../reducers/reducer'
import { SOCKETIO_URL } from '../constants/routes'

import { persistTargetPathReducer } from '../reducers/persistTargetPathReducer'

import CONFIG from '../config'
// import sessionTimeoutSaga from './sessionTimeout'
import authenticationStateSaga from './authenticationState'
import notificationsSaga from './notificationTrigger'
import socketAuthenticationSaga from './socketAuthentication'
import autocompleteInvalidatorSaga from './autocompleteInvalidator'
import { carbonIntensities } from '../actions/carbonIntensities'
import { defaultCarbonIntensities } from '../actions/defaultCarbonIntensities'
import { energyDensities } from '../actions/energyDensities'
import { energyEffectivenessRatios } from '../actions/energyEffectivenessRatios'
import { expectedUses } from '../actions/expectedUses'
import { fuelClasses } from '../actions/fuelClasses'
import { notionalTransferTypes } from '../actions/notionalTransferTypes'
import { petroleumCarbonIntensities } from '../actions/petroleumCarbonIntensities'
import { transactionTypes } from '../actions/transactionTypes'
import { roles } from '../actions/roleActions'
import autosaveSaga from './autosaveStore'
import { complianceReporting } from '../actions/complianceReporting'
import { exclusionReports } from '../actions/exclusionReports'

const sagaMiddleware = createSagaMiddleware()

const socket = io(SOCKETIO_URL)
const socketIoMiddleware = createSocketIoMiddleware(socket, 'socketio/')

const enhancer = compose(persistState(['targetPath'], { key: 'tfrs-state' }))

const combinedReducers = (state = {}, action) => {
  const currentRoute = state.routing || {}
  return {
    toastr: toastrReducer(state.toastr, action),
    targetPath: persistTargetPathReducer(state.targetPath, { ...action, currentRoute }),
    rootReducer: rootReducer(state.rootReducer, action)
  }
}

const allMiddleware = [
  socketIoMiddleware,
  sagaMiddleware,
  thunk
]

if (CONFIG.DEBUG.ENABLED) {
  allMiddleware.push(createLogger())
}

const store = configureStore({
  reducer: combinedReducers,
  middleware: [...allMiddleware],
  enhancers: [enhancer]
})

// sagaMiddleware.run(sessionTimeoutSaga); // TODO fix sessiontimeout with new keycloak login
sagaMiddleware.run(notificationsSaga, store)
sagaMiddleware.run(authenticationStateSaga, store)
sagaMiddleware.run(socketAuthenticationSaga, store)
sagaMiddleware.run(autocompleteInvalidatorSaga, store)
sagaMiddleware.run(autosaveSaga)

sagaMiddleware.run(roles.saga)
sagaMiddleware.run(complianceReporting.saga)
sagaMiddleware.run(exclusionReports.saga)
sagaMiddleware.run(carbonIntensities.saga)
sagaMiddleware.run(defaultCarbonIntensities.saga)
sagaMiddleware.run(energyDensities.saga)
sagaMiddleware.run(energyEffectivenessRatios.saga)
sagaMiddleware.run(expectedUses.saga)
sagaMiddleware.run(fuelClasses.saga)
sagaMiddleware.run(notionalTransferTypes.saga)
sagaMiddleware.run(petroleumCarbonIntensities.saga)
sagaMiddleware.run(transactionTypes.saga)

export default store
