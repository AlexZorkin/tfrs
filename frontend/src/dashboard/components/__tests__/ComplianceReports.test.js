import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import store from "../../../store/store";
import ComplianceReports from "../ComplianceReports";

describe("ComplianceReports", () => {
  test("Should render the component", () => {
    const props = {
      complianceReports: {
        isFinding: false,
        isGettingDashboard: false,
        items: [],
      },
      setFilter: jest.fn(),
    };
    const component = renderer.create(
      <BrowserRouter>
        <Provider store={store}>
          <ComplianceReports {...props} />
        </Provider>
      </BrowserRouter>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Checking the Analyst Count", () => {
    const props = {
      complianceReports: {
        isFinding: false,
        isGettingDashboard: false,
        items: [
          {
            id: 654,
            status: {
              analystStatus: "Unreviewed",
              directorStatus: "Unreviewed",
              fuelSupplierStatus: "Submitted",
              managerStatus: "Unreviewed",
            },
            type:'Compliance Report',
            supplementalReports:[]
          },
        ],
      },
      setFilter: jest.fn(),
    };
    const component = renderer.create(
      <BrowserRouter>
        <Provider store={store}>
          <ComplianceReports {...props} />
        </Provider>
      </BrowserRouter>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });  

  test("Checking the Manager Count", () => {
    const props = {
        complianceReports: {
            isFinding: false,
            isGettingDashboard: false,
            items: [
              {
                id: 294,
                status: {
                  analystStatus: "Unreviewed",
                  directorStatus: "Unreviewed",
                  managerStatus: "Unreviewed",
                },
                type:'Compliance Report',
                supplementalReports:[{
                    status: {
                        analystStatus: "Recommended",
                        directorStatus: "Unreviewed",
                        managerStatus: "Unreviewed",
                    } 
                },]
              },{
                id: 295,
                status: {},
                type:'Exclusion Report',
                supplementalReports:[{
                    status: {},
                    supplementalReports:[{
                        status: {
                            analystStatus: "Recommended",
                            directorStatus: "Unreviewed",
                            managerStatus: "Unreviewed",
                        } 
                    },]
                },]
              },
            ],
          },
          setFilter: jest.fn(),
    }
    const component = renderer.create(
        <BrowserRouter>
            <Provider store={store}>
                <ComplianceReports {...props}/>
            </Provider>
        </BrowserRouter>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    
  });

  test("Checking the Director Count", () => {
    const props = {
      complianceReports: {
        isFinding: false,
        isGettingDashboard: false,
        items: [
          {
            id: 294,
            status: {
              analystStatus: "Unreviewed",
              directorStatus: "Unreviewed",
              managerStatus: "Unreviewed",
            },
            type:'Compliance Report',
            supplementalReports:[{
                status: {
                    analystStatus: "Recommended'",
                    directorStatus: "Unreviewed",
                    managerStatus: "Not Recommended",
                } 
            }]
          },
        ],
      },
      setFilter: jest.fn(),
    };
    const component = renderer.create(
      <BrowserRouter>
        <Provider store={store}>
          <ComplianceReports {...props} />
        </Provider>
      </BrowserRouter>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
