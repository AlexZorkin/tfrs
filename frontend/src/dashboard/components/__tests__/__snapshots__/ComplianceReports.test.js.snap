// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComplianceReports Checking the Analyst Count 1`] = `
<div
  className="dashboard-fieldset"
>
  <h1>
    Compliance & Exclusion Reports
  </h1>
  There are:
  <div>
    <div
      className="value"
    >
      1
    </div>
    <div
      className="content"
    >
      <h2>
        compliance reports in progress:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          1
           awaiting government analyst review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting compliance manager review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting Director review
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="value"
    >
      0
    </div>
    <div
      className="content"
    >
      <h2>
        exclusion reports in progress:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting government analyst review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting compliance manager review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting Director review
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="content"
    >
      <h2>
        View all reports:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          Current compliance period
        </button>
         | 
        <button
          onClick={[Function]}
          type="button"
        >
          All/historical
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ComplianceReports Checking the Director Count 1`] = `
<div
  className="dashboard-fieldset"
>
  <h1>
    Compliance & Exclusion Reports
  </h1>
  There are:
  <div>
    <div
      className="value"
    >
      1
    </div>
    <div
      className="content"
    >
      <h2>
        compliance reports in progress:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting government analyst review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting compliance manager review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          1
           awaiting Director review
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="value"
    >
      0
    </div>
    <div
      className="content"
    >
      <h2>
        exclusion reports in progress:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting government analyst review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting compliance manager review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting Director review
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="content"
    >
      <h2>
        View all reports:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          Current compliance period
        </button>
         | 
        <button
          onClick={[Function]}
          type="button"
        >
          All/historical
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ComplianceReports Checking the Manager Count 1`] = `
<div
  className="dashboard-fieldset"
>
  <h1>
    Compliance & Exclusion Reports
  </h1>
  There are:
  <div>
    <div
      className="value"
    >
      1
    </div>
    <div
      className="content"
    >
      <h2>
        compliance reports in progress:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting government analyst review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          1
           awaiting compliance manager review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting Director review
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="value"
    >
      1
    </div>
    <div
      className="content"
    >
      <h2>
        exclusion reports in progress:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting government analyst review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          1
           awaiting compliance manager review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting Director review
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="content"
    >
      <h2>
        View all reports:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          Current compliance period
        </button>
         | 
        <button
          onClick={[Function]}
          type="button"
        >
          All/historical
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ComplianceReports Should render the component 1`] = `
<div
  className="dashboard-fieldset"
>
  <h1>
    Compliance & Exclusion Reports
  </h1>
  There are:
  <div>
    <div
      className="value"
    >
      0
    </div>
    <div
      className="content"
    >
      <h2>
        compliance reports in progress:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting government analyst review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting compliance manager review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting Director review
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="value"
    >
      0
    </div>
    <div
      className="content"
    >
      <h2>
        exclusion reports in progress:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting government analyst review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting compliance manager review
        </button>
      </div>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          0
           awaiting Director review
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="content"
    >
      <h2>
        View all reports:
      </h2>
      <div>
        <button
          onClick={[Function]}
          type="button"
        >
          Current compliance period
        </button>
         | 
        <button
          onClick={[Function]}
          type="button"
        >
          All/historical
        </button>
      </div>
    </div>
  </div>
</div>
`;
